goldenmaster/apigear/mqtt/mqttcppclient.cpp

```diff
@@ -0,0 +1,578 @@
+            InvokeReplyFunc func {};
+            m_invokeResultsMutex.lock();
+            if((m_invokeResults.find(randomId) != m_invokeResults.end()))
+            {
+                func = m_invokeResults[randomId];
+                m_invokeResults.erase(randomId);
+            }
+            m_invokeResultsMutex.unlock();
+            if(func) {
+                const InvokeReplyArg arg{nlohmann::json::parse(message.content)};
+                func(arg);
+            }
+        }
+    }
+}
+
+void Client::invokeRemote(const Topic& topic, const std::string& value, InvokeReplyFunc func)
+{
+    const std::string responseTopic = topic.getEncodedTopic() + "/" + m_clientID + "/result" ;
+
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    int rc = 0;
+
+    MQTTProperty responseTopicProperty;
+    responseTopicProperty.identifier = MQTTPROPERTY_CODE_RESPONSE_TOPIC;
+    responseTopicProperty.value.data = { static_cast<int>(responseTopic.size()), const_cast<char*>(responseTopic.c_str()) };
+    MQTTProperties_add(&(opts.properties), &responseTopicProperty);
+
+    MQTTProperty correlationDataProperty;
+    correlationDataProperty.identifier = MQTTPROPERTY_CODE_CORRELATION_DATA;
+    std::uniform_int_distribution<> distribution (0, 100000);
+
+    auto randomId = 0;
+    m_invokeResultsMutex.lock();
+    do {
+        randomId = distribution(randomNumberGenerator);
+    } while (m_invokeResults.find(randomId) != m_invokeResults.end());
+    m_invokeResults.insert(std::pair<int, InvokeReplyFunc>(randomId, func));
+    m_invokeResultsMutex.unlock();
+    const std::string correlationData = { std::to_string(randomId) } ;
+    correlationDataProperty.value.data = { static_cast<int>(correlationData.size()), const_cast<char*>(correlationData.c_str()) };
+    MQTTProperties_add(&(opts.properties), &correlationDataProperty);
+
+    std::cout << "corrdata " << correlationData << std::endl;
+    opts.onSuccess5 = onSendSuccess;
+    opts.onFailure5 = onSendFailure;
+    opts.context = this;
+    // opts.properties = properties;
+    pubmsg.payload = const_cast<void*>(static_cast<const void*>(value.c_str()));
+    pubmsg.payloadlen = value.size();
+    pubmsg.qos = QOS;
+    pubmsg.retained = 0;
+
+    // the responseOptions properties do get overwritten by the msg properties later
+    pubmsg.properties = opts.properties;
+
+    if ((rc = MQTTAsync_sendMessage(*m_client.get(), topic.getEncodedTopic().c_str(), &pubmsg, &opts)) != MQTTASYNC_SUCCESS)
+    {
+        printf("Failed to start sendMessage, return code %d\n", rc);
+    }
+}
+
+void Client::notifyPropertyChange(const Topic& topic, const std::string& value)
+{
+    MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
+    MQTTAsync_message pubmsg = MQTTAsync_message_initializer;
+    int rc = 0;
+
+    opts.onSuccess5 = onSendSuccess;
```

Comment:
I thikn the opts could be created only once for the lifetime of client not on each property change, you could maybe create some sender? and pass there only the part that changes (pubmsg)
In general, sending a message is on hot path and should be as fastest as possible

Reply:
removed the call since we didn't really use it.
In the future we should come up with a strategy for messages which couldn't be sent.
Then we need to use onSuccess/onFailure

Reply:
https://github.com/apigear-io/template-cpp14/issues/109